controllers/fluxcd/application-controller_test.go

```diff
@@ -0,0 +1,683 @@
+/*
+Copyright 2022 The KubeSphere Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fluxcd
+
+import (
+	"context"
+	"github.com/stretchr/testify/assert"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	apischema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"kubesphere.io/devops/pkg/api/gitops/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"testing"
+	"time"
+)
+
+func TestApplicationReconciler_Reconcile(t *testing.T) {
+	schema, err := v1alpha1.SchemeBuilder.Register().Build()
+	assert.Nil(t, err)
+
+	argoApp := &v1alpha1.Application{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-app",
+			Namespace: "fake-ns",
+		},
+		Spec: v1alpha1.ApplicationSpec{
+			Kind: v1alpha1.ArgoCD,
+		},
+	}
+
+	fluxApp := argoApp.DeepCopy()
+	fluxApp.Spec.Kind = v1alpha1.FluxCD
+
+	type fields struct {
+		Client client.Client
+	}
+	type args struct {
+		req ctrl.Request
+	}
+
+	tests := []struct {
+		name       string
+		fields     fields
+		args       args
+		wantResult ctrl.Result
+		wantErr    assert.ErrorAssertionFunc
+	}{
+		{
+			name: "not found an application",
+			fields: fields{
+				Client: fake.NewFakeClientWithScheme(schema),
+			},
+			args: args{
+				req: ctrl.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: "fake-app",
+						Name:      "fake-ns",
+					},
+				},
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name: "found an argocd application",
+			fields: fields{
+				Client: fake.NewFakeClientWithScheme(schema, argoApp.DeepCopy()),
+			},
+			args: args{
+				req: ctrl.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: "fake-app",
+						Name:      "fake-ns",
+					},
+				},
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name: "found a fluxcd application",
+			fields: fields{
+				Client: fake.NewFakeClientWithScheme(schema, fluxApp.DeepCopy()),
+			},
+			args: args{
+				req: ctrl.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: "fake-ns",
+						Name:      "fake-app",
+					},
+				},
+			},
+			wantErr: assert.NoError,
+		},
```

Comment:
I actually wrote unit test for this but these test case has not find this bug. @LinuxSuRen

Reply:
Yes, it would be great if we could cover that case in unit tests.